sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,916 @@
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 2 ); gens = f.generators
+        >>> a, b = gens[0], gens[1]; w = a**5*b*a**2*b**-4*a;
+        >>> w
+        a**5*b*a**2*b**-4*a
+        >>> len( w )
+        13
+        >>> len( a**17 )
+        17
+        >>> len( w**0 )
+        0
+
+        """
+        return sum([abs(syllable[1]) for syllable in self.array_form])
+
+    def inverse(self):
+        """
+        Returns the inverse of a `FreeGroupElm` element
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 2, "swapnil" )
+        >>> f[0].inverse()
+        swapnil0**-1
+        >>> (f[0]*f[1]).inverse()
+        swapnil1**-1*swapnil0**-1
+
+        """
+        new_array = [(i, -j) for i, j in self.array_form[::-1]]
+        return FreeGroupElm(self.group, new_array, self.str_expr)
+
+    def order(self):
+        """Find the order of a `FreeGroupElm`.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> (f[0]**2*f[1]*f[1]**-1*f[0]**-2).order()
+        1
+
+        """
+        if self.is_Identity:
+            return 1
+        else:
+            return S.Infinity
+
+    def commutator(self, other):
+        """Returns the commutator of self and x: ``~x*~self*x*self``
+        """
+        if not isinstance(other, FreeGroupElm) or self.group != other.group:
+            raise ValueError("commutator of only `FreeGroupElm` of the same "
+                    "`FreeGroup` exists")
+        else:
+            return self.inverse()*other.inverse()*self*other
+
+    def eliminate_word(self, gen, by):
+        """
+        For an associative word `self`, a generator `gen`, and an associative word
+        by, `eliminate_word` returns the associative word obtained by replacing
+        each occurrence of `gen` in `self` by `by`.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> w = f[0]**5*f[1]*f[0]**2*f[1]**-4*f[0]
+        >>> w.eliminate_word( f[0], f[0]**2 )
+        f0**10*f1*f0**4*f1**-4*f0**2
+
+        >>> w.eliminate_word( f[0], f[1]**-1 )
+        f1**-11
+
+        """
+        e = self.ext_rep()
+        gen = gen.generator_syllable(0)
+        l = []
+        for i in range(0, len(e) - 1, 2):
+            if e[i] == gen:
+                app = (by**e[i + 1]).ext_rep()
+            else:
+                app = e[i: i + 2]
+            j = len(l) - 1
+            while j > 0 and len(app) > 0 and l[j - 1] == app[0]:
+                s = l[j] + app[1]
+                if s == 0:
+                    j = j - 2
+                else:
+                    l[j] = s
+                app = app[2: len(app)]
+
+            if j + 1 < len(l):
+                l = l[0: j + 1]
+
+            if len(app) > 0:
+                l.append(tuple(app))
+        zero_simp(l)
+        mult_simp(l)
+        zero_simp(l)
```

Comment:
After the second `zero_simp(l)` another `mult_simp(l)` may be needed.  Perhaps those functions should be combined.